include/nbl/video/IDescriptorPool.h

```diff
@@ -32,86 +30,186 @@ class NBL_API IDescriptorPool : public core::IReferenceCounted, public IBackendO
-            m_UBO_SSBOStorage = std::make_unique<core::StorageTrivializer<core::smart_refctd_ptr<IGPUBuffer>>[]>(m_maxDescriptorCount[asset::EDT_UNIFORM_BUFFER] + m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER] + m_maxDescriptorCount[asset::EDT_UNIFORM_BUFFER_DYNAMIC] + m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER_DYNAMIC]);
-            m_UTB_STBStorage = std::make_unique<core::StorageTrivializer<core::smart_refctd_ptr<IGPUBufferView>>[]>(m_maxDescriptorCount[asset::EDT_UNIFORM_TEXEL_BUFFER] + m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER_DYNAMIC]);
-            m_accelerationStructureStorage = std::make_unique<core::StorageTrivializer<core::smart_refctd_ptr<IGPUAccelerationStructure>>[]>(m_maxDescriptorCount[asset::EDT_ACCELERATION_STRUCTURE]);
-        }
-
-        virtual ~IDescriptorPool() {}
-
-        virtual core::smart_refctd_ptr<IGPUDescriptorSet> createDescriptorSet_impl(core::smart_refctd_ptr<const IGPUDescriptorSetLayout>&& layout) = 0;
-        virtual bool freeDescriptorSets_impl(const uint32_t descriptorSetCount, IGPUDescriptorSet* const* const descriptorSets) = 0;
+            inline void reset(const bool allowsFreeing)
+            {
+                if (!allowsFreeing)
+                    linearAllocator.reset();
+                else
+                    generalAllocator.reset();
+            }
 
-        uint32_t m_maxSets;
+            inline uint32_t getAllocatedDescriptorCount(const bool allowsFreeing) const
+            {
+                if (!allowsFreeing)
+                    return linearAllocator.get_allocated_size();
+                else
+                    return generalAllocator.get_allocated_size();
+            }
 
-    private:
-        const IDescriptorPool::E_CREATE_FLAGS m_flags;
-        uint32_t m_maxDescriptorCount[asset::EDT_COUNT];
-        union
-        {
-            core::LinearAddressAllocator<uint32_t> m_linearAllocators[asset::EDT_COUNT+1];
-            core::GeneralpurposeAddressAllocator<uint32_t> m_generalAllocators[asset::EDT_COUNT+1];
+            union
+            {
+                core::LinearAddressAllocator<uint32_t> linearAllocator;
+                core::GeneralpurposeAddressAllocator<uint32_t> generalAllocator;
+            };
+            std::unique_ptr<uint8_t[]> generalAllocatorReservedSpace = nullptr;
         };
-        std::unique_ptr<uint8_t[]> m_generalAllocatorReservedSpace[asset::EDT_COUNT+1];
+        std::unique_ptr<allocator_state_t> m_descriptorAllocators[static_cast<uint32_t>(asset::IDescriptor::E_TYPE::ET_COUNT) + 1];
+
+        const SCreateInfo m_creationParameters;
+        std::atomic_uint32_t m_version;
 
         std::unique_ptr<core::StorageTrivializer<core::smart_refctd_ptr<video::IGPUImageView>>[]> m_textureStorage;
         std::unique_ptr<core::StorageTrivializer<core::smart_refctd_ptr<video::IGPUSampler>>[]> m_mutableSamplerStorage;
```

Comment:
ok we don't want to trivialize the storage, make these `std::unique_ptr<core::smart_refctd_ptr<DESCRITPOR_TYPE>[]>` instead.

It causes a bunch of problems like: https://github.com/Devsh-Graphics-Programming/Nabla/pull/345/files/8fbc87079edc5109a9c916eeb7ea7c0eff9ce128..b3719ddb1377d8af8f8d5f4dd112d3c844394553#r1116336481

> hmm maybe the reason why I didn't want to placement new an array, is that I was assuming very very very large descriptor pools and not having to iterate over all of the items to create/destroy them.
But this assumption falls apart with the pool allocator, cause the ranges can be vastly disjoint and "live" handles can be anywhere.

> And anyway even if we had a 2M descriptor pool, to visit every smart pointer its 16MB, even with 20GB/s single channel DDR4, you're looking at 781 us to reset (and initialize) such a large pool.

Reply:
ok actually, almost 1ms for a reset or free sounds pretty bad.  

If you do this https://github.com/Devsh-Graphics-Programming/Nabla/pull/345/files/8fbc87079edc5109a9c916eeb7ea7c0eff9ce128..b3719ddb1377d8af8f8d5f4dd112d3c844394553#r1116359755
and keep track of all currently allocated descriptor sets, then you could keep the storage trivialized.

But you'd need to:
1. Assert in `~IDescriptorPool` that there are no currently allocated DescriptorSets (impossibru since DS holds a back smartpointer to the Pool)
2. construct the storage range as part of offset allocation
3. destruct the storage range as part of `reset` or individual DescriptorSet free

Reply:
> construct the storage range as part of offset allocation

Why do you think it is important to construct the storage range as part of offset allocation as opposed to do the construction in `IGPUDescriptorSet`, [like what is happening now](https://github.com/Devsh-Graphics-Programming/Nabla/blob/1cd208bb98bdaa9f3eaef5530458ec23aeb2861b/src/nbl/video/IGPUDescriptorSet.cpp#L21)?

Reply:
I was originally thinking about keeping the `default_construct` where the `destroy_n` is for clarity, but I see the benefit of doing it in the `IGPUDescriptorSet` in-case there are any failures and also to keep the pool-mutex locked as little as possible.

Edited the TODO list.

Reply:
For 1. see this comment https://github.com/Devsh-Graphics-Programming/Nabla/pull/345#discussion_r1121143139

Reply:
Resolved?